subsys/net/lib/lwm2m/lwm2m_obj_firmware.c

```diff
@@ -59,33 +59,185 @@ static struct lwm2m_engine_obj_inst inst;
+		}
+
+		lwm2m_firmware_set_update_state(STATE_IDLE);
+		break;
+	case RESULT_NO_STORAGE:
+	case RESULT_OUT_OF_MEM:
+	case RESULT_CONNECTION_LOST:
+	case RESULT_UNSUP_FW:
+	case RESULT_INVALID_URI:
+	case RESULT_UNSUP_PROTO:
+		if (update_state != STATE_DOWNLOADING) {
+			error = true;
+			state = update_state;
+		}
+
+		lwm2m_firmware_set_update_state(STATE_IDLE);
+		break;
+	case RESULT_INTEGRITY_FAILED:
+		if (update_state != STATE_DOWNLOADING &&
+		    update_state != STATE_UPDATING) {
+			error = true;
+			state = update_state;
+		}
+
+		lwm2m_firmware_set_update_state(STATE_IDLE);
+		break;
+	case RESULT_UPDATE_FAILED:
+		if (update_state != STATE_UPDATING) {
+			error = true;
+			state = update_state;
+		}
+
+		/* Next state could be idle or downloaded */
+		break;
+	default:
+		SYS_LOG_ERR("Unhandled result: %u", result);
+		return;
+	}
+
+	if (error) {
+		SYS_LOG_ERR("Unexpected result(%u) set while state is %u",
+			    result, state);
+	}
+
+	update_result = result;
+	NOTIFY_OBSERVER(LWM2M_OBJECT_FIRMWARE_ID, 0, FIRMWARE_UPDATE_RESULT_ID);
+	SYS_LOG_DBG("Update result = %d", update_result);
+}
+
 static int package_write_cb(u16_t obj_inst_id,
 			    u8_t *data, u16_t data_len,
 			    bool last_block, size_t total_size)
 {
-	SYS_LOG_DBG("PACKAGE WRITE");
+	u8_t state;
+	int ret = 0;
+
+	state = lwm2m_firmware_get_update_state();
+	if (state == STATE_IDLE) {
+		/* TODO: setup timer to check download status,
+		 * make sure it fail after timeout
+		 */
+		lwm2m_firmware_set_update_state(STATE_DOWNLOADING);
+	} else if (state != STATE_DOWNLOADING) {
+		if (data_len == 0 && state == STATE_DOWNLOADED) {
+			/* reset to state idle and result default */
+			lwm2m_firmware_set_update_result(RESULT_DEFAULT);
+			return 1;
+		}
+
+		SYS_LOG_DBG("Cannot download: state = %d", state);
+		return -EPERM;
+	}
+
 	if (write_cb) {
-		write_cb(obj_inst_id, data, data_len, last_block, total_size);
-		return 1;
+		ret = write_cb(obj_inst_id, data, data_len,
+			       last_block, total_size);
+		if (ret < 0) {
+			SYS_LOG_ERR("Failed to store firmware: %d", ret);
+			lwm2m_firmware_set_update_result(
+					RESULT_INTEGRITY_FAILED);
+		}
 	}
 
-	return 0;
+	if (last_block) {
+		lwm2m_firmware_set_update_state(STATE_DOWNLOADED);
+	}
+
+	return 1;
 }
 
 static int package_uri_write_cb(u16_t obj_inst_id,
 				u8_t *data, u16_t data_len,
 				bool last_block, size_t total_size)
 {
 	SYS_LOG_DBG("PACKAGE_URI WRITE: %s", package_uri);
-#ifdef CONFIG_LWM2M_FIRMWARE_UPDATE_PULL_SUPPORT
-	lwm2m_firmware_start_transfer(package_uri);
-	return 1;
-#endif
+
+	if (IS_ENABLED(CONFIG_LWM2M_FIRMWARE_UPDATE_PULL_SUPPORT)) {
+		u8_t state = lwm2m_firmware_get_update_state();
+
+		if (state == STATE_IDLE) {
+			lwm2m_firmware_set_update_result(RESULT_DEFAULT);
+			lwm2m_firmware_start_transfer(package_uri);
+		} else if (state == STATE_DOWNLOADED && data_len == 0) {
+			/* reset to state idle and result default */
+			lwm2m_firmware_set_update_result(RESULT_DEFAULT);
+		}
+
```

Comment:
Mind also adding a error message here in case this gets called when state == DOWNLOADING.

Reply:
Will print out a error message here.